File: cpp14/introspection.py
class A(object):
  # simply overrides the 'object.__str__' method
  def __str__(self):
    return "I'm an A!"

class B(object):
  # a custom method for my custom objects that I want to serialize
  def serialize(self):
    return "I'm a  B!"

class C(object):
  def __init__(self):
    # oops! 'serialize' is not a method
    self.serialize = 0

  def __str__(self):
    return "I'm a  C!"

def serialize(obj):
  # check if obj has an attribute called 'serialize'
  if hasattr(obj, "serialize"):
    # let's check if that 'serialize' attribute is a method
    if hasattr(obj.serialize, "__call__"):
      return obj.serialize()
  # else we call the __str__ method
  return str(obj)


if __name__ == "__main__":
  a = A()
  b = B()
  c = C()

  print(serialize(a))
  print(serialize(b))
  print(serialize(c))

File: cpp14/introspection.cpp
#include <boost/hana.hpp>
#include <iostream>
#include <string>

namespace hana = boost::hana;

// check if a type has a serialize method
auto hasSerialize = hana::is_valid([](auto&& x) -> decltype(x.serialize()) { });

// serialize any kind of objects
template<typename T>
std::string serialize(const T& obj) {
  return hana::if_(hasSerialize(obj), [](auto& x) {return x.serialize();}, [](auto& x) {return to_string(x);})(obj);
}

// type A with only a to_string overload
struct A {};

std::string to_string(const A&) {
  return "I'm an A!";
}

// type B with a serialize method
struct B {
  std::string serialize() const {
    return "I'm a  B!";
  }
};

// type C with a "wrong" serialize member (not a method) and a to_string overload
struct C {
  std::string serialize;
};

std::string to_string(const C&) {
  return "I'm a  C!";
}

int main() {
  A a;
  B b;
  C c;

  std::cout << serialize(a) << std::endl;
  std::cout << serialize(b) << std::endl;
  std::cout << serialize(c) << std::endl;
}
